File: 564_nearest_palindromic/nearest_palindromic.cpp
/*
 * @Date: 2022-03-02 00:10:32
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2022-03-02 00:41:39
 * @FilePath: /algorithm/564_nearest_palindromic/nearest_palindromic.cpp
 */

#include <cassert>
#include <cmath>
#include <string>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<long> getCandidates(const string& n) {
    int len = n.length();
    vector<long> candidates = {
        (long)pow(10, len - 1) - 1,
        (long)pow(10, len) + 1,
    };
    long selfPrefix = stol(n.substr(0, (len + 1) / 2));
    for (int i : {selfPrefix - 1, selfPrefix, selfPrefix + 1}) {
      string prefix = to_string(i);
      string candidate =
          prefix + string(prefix.rbegin() + (len & 1), prefix.rend());
      candidates.push_back(stol(candidate));
    }
    return candidates;
  }

  string nearestPalindromic(string n) {
    long selfNumber = stol(n), ans = -1;
    const vector<long>& candidates = getCandidates(n);
    for (auto& candidate : candidates) {
      if (candidate != selfNumber) {
        if (ans == -1 || abs(candidate - selfNumber) < abs(ans - selfNumber) ||
            (abs(candidate - selfNumber) == abs(ans - selfNumber) &&
             candidate < ans)) {
          ans = candidate;
        }
      }
    }
    return to_string(ans);
  }
};

int main() {
  assert(Solution().nearestPalindromic("123") == "121");
  assert(Solution().nearestPalindromic("1") == "0");
}

File: 564_nearest_palindromic/nearest_palindromic.go
/*
 * @Date: 2022-03-02 00:10:35
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2022-03-02 01:29:08
 * @FilePath: /algorithm/564_nearest_palindromic/nearest_palindromic.go
 */

package main

import (
	"math"
	"strconv"
)

func nearestPalindromic(n string) string {
	abs := func(x int) int {
		if x < 0 {
			return -x
		}
		return x
	}
	m := len(n)
	candidates := []int{int(math.Pow10(m-1)) - 1, int(math.Pow10(m)) + 1}
	selfPrefix, _ := strconv.Atoi(n[:(m+1)/2])
	for _, x := range []int{selfPrefix - 1, selfPrefix, selfPrefix + 1} {
		y := x
		if m&1 == 1 {
			y /= 10
		}
		for ; y > 0; y /= 10 {
			x = x*10 + y%10
		}
		candidates = append(candidates, x)
	}

	ans := -1
	selfNumber, _ := strconv.Atoi(n)
	for _, candidate := range candidates {
		if candidate != selfNumber {
			if ans == -1 ||
				abs(candidate-selfNumber) < abs(ans-selfNumber) ||
				abs(candidate-selfNumber) == abs(ans-selfNumber) && candidate < ans {
				ans = candidate
			}
		}
	}
	return strconv.Itoa(ans)
}

func main() {
	assert := func(b bool) {
		if !b {
			panic("Not Passed")
		}
	}
	assert(nearestPalindromic("123") == "121")
	assert(nearestPalindromic("1") == "0")
}
